packages/theme/src/cli/utilities/asset-ignore.ts

```diff
@@ -56,14 +61,31 @@ function matchGlob(key: string, pattern: string) {
   // When the the standard match fails and the pattern includes '/*.', we
   // replace '/*.' with '/**/*.' to emulate Shopify CLI 2.x behavior, as it was
   // based on 'File.fnmatch'.
-  if (pattern.includes('/*.')) {
-    outputWarn(`The pattern ${pattern} is unsafe, please use a valid regex or glob.`)
+  if (shouldReplaceGlobPattern(pattern)) {
     return originalMatchGlob(key, pattern.replace('/*.', '/**/*.'))
   }
 
   return false
 }
 
+function raiseWarningForNonExplicitGlobPatterns(patterns: string[]) {
+  const allPatterns = new Set(patterns)
+  allPatterns.forEach((pattern) => {
+    if (shouldReplaceGlobPattern(pattern)) {
+      outputWarn(
+        `Warning: The pattern '${pattern}' does not include subdirectories. To maintain backwards compatibility, we have modified your pattern to ${pattern.replace(
+          '/*.',
+          '/**/*.',
+        )} to explicitly include subdirectories.`,
+      )
+    }
+  })
+}
+
+function shouldReplaceGlobPattern(pattern: string): boolean {
```

Comment:
Decision: I chose to limit this to the `templates` folder because that's the only folder where [subdirectories are supported](https://shopify.dev/docs/themes/architecture#directory-structure-and-component-types).

Some questions that could influence this
- Do you consider something like `assets/*.liquid` to be unsafe / something we need to substitute?
  - If we recommend `assets/**/*.liquid`, that could be confusing since there are no subdirectories there 

- Will we remove the automatic conversion of the glob pattern one day so that `templates/*.json` and `templates/**/*.json` mean different things in the future?
  - Some context [here](https://github.com/Shopify/cli/issues/3664)

Reply:
+1 for restricting this warning to templates, as it's the only directory that may have a sub-directory :)

Regarding the removal of this warning in the future, because it would be a destructive breaking change (as in some contexts it may imply file deletion), I'd suggest making this change in the next major release